prov/tcp/src/tcpx_comm.c

```diff
@@ -36,135 +36,122 @@
 #include <ofi_util.h>
 #include "tcpx.h"
 
-static ssize_t tcpx_comm_recv_socket(SOCKET conn_fd, void *buf, size_t len)
+static void prepare_iovec(struct iovec *iovec,
+			  struct tcpx_pe_entry *pe_entry)
 {
-	ssize_t ret;
-
-	ret = ofi_recv_socket(conn_fd, buf, len, 0);
-	if (ret > 0) {
-		FI_DBG(&tcpx_prov, FI_LOG_EP_DATA,
-		       "read from network: %lu\n", ret);
-	} else {
-		if (ret < 0) {
-			FI_DBG(&tcpx_prov, FI_LOG_EP_DATA,
-			       "read %s\n", strerror(ofi_sockerr()));
-			ret = 0;
-		} else {
-			FI_DBG(&tcpx_prov, FI_LOG_EP_DATA," socket closed\n");
-			return ret;
-		}
+	int i = 0;
+
+	iovec[0].iov_base = (void *) &pe_entry->msg_hdr;
+	iovec[0].iov_len = sizeof(pe_entry->msg_hdr);
+
+	for (i = 0 ; i < pe_entry->msg_data.iov_cnt ; i++) {
+		iovec[i+1].iov_base = (void *) pe_entry->msg_data.iov[i].addr;
+		iovec[i+1].iov_len = pe_entry->msg_data.iov[i].len;
 	}
-	return ret;
-}
 
-static void tcpx_comm_recv_buffer(struct tcpx_pe_entry *pe_entry)
-{
-	int ret;
-	size_t max_read, avail;
-
-	avail = ofi_rbavail(&pe_entry->comm_buf);
-	assert(avail == pe_entry->comm_buf.size);
-	pe_entry->comm_buf.rcnt =
-		pe_entry->comm_buf.wcnt =
-		pe_entry->comm_buf.wpos = 0;
-
-	max_read = pe_entry->msg_hdr.size - pe_entry->done_len;
-	ret = tcpx_comm_recv_socket(pe_entry->ep->conn_fd, (char *) pe_entry->comm_buf.buf,
-				    MIN(max_read, avail));
-	pe_entry->comm_buf.wpos += ret;
-	ofi_rbcommit(&pe_entry->comm_buf);
 }
-
-ssize_t tcpx_comm_recv(struct tcpx_pe_entry *pe_entry, void *buf, size_t len)
+static void prepare_msg(struct msghdr *msghdr, struct iovec *iovec,
+			struct tcpx_pe_entry *pe_entry)
 {
-	ssize_t read_len;
-	if (ofi_rbempty(&pe_entry->comm_buf)) {
-		if (len <= pe_entry->cache_sz) {
-			tcpx_comm_recv_buffer(pe_entry);
-		} else {
-			return tcpx_comm_recv_socket(pe_entry->ep->conn_fd, buf, len);
-		}
+	size_t len = 0;
+	uint64_t buf_offset;
+	int iov_offset, i;
+
+	prepare_iovec(iovec, pe_entry);
+
+	iov_offset = 0;
+	buf_offset = pe_entry->done_len;
+	for (i = 0 ; i < pe_entry->msg_data.iov_cnt+1 ; i++) {
+		len += iovec[i].iov_len;
+		if (pe_entry->done_len < len)
+			continue;
+
+		iov_offset = i+1;
+		buf_offset = pe_entry->done_len - len;
 	}
 
-	read_len = MIN(len, ofi_rbused(&pe_entry->comm_buf));
-	ofi_rbread(&pe_entry->comm_buf, buf, read_len);
-	FI_DBG(&tcpx_prov, FI_LOG_EP_DATA, "read from buffer: %lu\n", read_len);
-	return read_len;
+	iovec[iov_offset].iov_base = ((uint8_t *)(iovec[iov_offset].iov_base) +
+				      buf_offset);
+
+	iovec[iov_offset].iov_len -= buf_offset;
+
+	msghdr->msg_name = NULL;
+	msghdr->msg_namelen = 0;
+	msghdr->msg_iov = &iovec[iov_offset];
+	msghdr->msg_iovlen = pe_entry->msg_data.iov_cnt+1 - iov_offset;
+	msghdr->msg_control = NULL;
+	msghdr->msg_controllen = 0;
+	msghdr->msg_flags = 0;
```

Comment:
Most the msghdr fields are being set to 0.  Maybe just use writev() instead of sendmsg()?